### PR TITLE
feat(es/minfiier): Compute more with sequential inliner

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

**Related issue:**

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_transform_optmization
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/tests/terser/compress/arrays/index/output.js [hunk 1]
```diff
@@ -1 +1,5 @@
+var a = [
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/terser/compress/arrays/index/output.js:1]: This seems like a regression

**Review by jridgewell — DISMISSED** [review]

**Review by jridgewell — DISMISSED** [review]

File: crates/swc_ecma_minifier/src/compress/optimize/sequences.rs [hunk 2]
```diff
@@ -2411,3 +2439,12 @@ pub(crate) fn is_trivial_lit(e: &Expr) -> bool {
         _ => false,
     }
 }
+
+/// This assumes `a.left.to_id() == b.left.to_id()`
+fn can_drop_op_for(a: AssignOp, b: AssignOp) -> bool {
+    if a == b {
+        return matches!(a, op!("+=") | op!("-=") | op!("*="));
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/compress/optimize/sequences.rs:2446]: I'll improve this function with another PR

File: crates/swc_ecma_minifier/tests/pass-1/compute/1/output.js [hunk 3]
```diff
@@ -1,4 +1,9 @@
 console.log((function() {
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/pass-1/compute/1/output.js:1]: This will be improved by another PR

File: crates/swc_ecma_minifier/tests/fixture/next/33265/static/chunks/pages/index-cb36c1bf7f830e3c/output.js [hunk 4]
```diff
@@ -68,7 +68,7 @@
                 }, BigInt(0));
                 if (void 0 !== _ref$signed && _ref$signed) {
                     var max = BYTE_TABLE[bytes.length] / BigInt(2) - BigInt(1);
-                    (number = BigInt(number)) > max && (number -= max, number -= max, number -= BigInt(2));
+                    (number = BigInt(number)) > max && (number = max - max - BigInt(2));
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/fixture/next/33265/static/chunks/pages/index-cb36c1bf7f830e3c/output.js:71]: This seems wrong

**Comment by kdy1** [comment]
@jridgewell I'm sorry, but can you review again when you have time?

File: crates/swc/tests/tsc-references/destructuringControlFlow.2.minified.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
 //// [destructuringControlFlow.ts]
-(0, [
+[
     "foo"
-][1]).toUpperCase();
+][1].toUpperCase();
```
Comment (jridgewell) [inline crates/swc/tests/tsc-references/destructuringControlFlow.2.minified.js:4]: Not that it matters in this specific case, but this has changed the `this` context of the method.

File: crates/swc_ecma_minifier/src/compress/optimize/sequences.rs [hunk 6]
```diff
@@ -2414,3 +2449,16 @@ pub(crate) fn is_trivial_lit(e: &Expr) -> bool {
...
 }
+
+/// This assumes `a.left.to_id() == b.left.to_id()`
+fn can_drop_op_for(a: AssignOp, b: AssignOp) -> bool {
+    if a == op!("=") {
+        return true;
+    }
+
+    if a == b {
+        return matches!(a, op!("+=") | op!("*="));
```
Comment (jridgewell) [inline crates/swc_ecma_minifier/src/compress/optimize/sequences.rs:2460]: This is just looking for associative property? We should be able to merge `/=`, etc?

File: crates/swc/tests/tsc-references/destructuringControlFlow.2.minified.js (hunk 5 above)
Comment (kdy1) [reply crates/swc/tests/tsc-references/destructuringControlFlow.2.minified.js:4]: Nice catch, I'll add a test case for it.

File: crates/swc_ecma_minifier/src/compress/optimize/sequences.rs (hunk 6 above)
Comment (kdy1) [reply crates/swc_ecma_minifier/src/compress/optimize/sequences.rs:2460]: This first one should be prefixed with `-` so I just excluded it

**Review by jridgewell — DISMISSED** [review]
